게임 기본 로직 제작

GyutaeLee opened the pull request:
1. 게임 기본 상태 구현 : time, ui
2. 유닛 기본 능력치 클래스 구현
3. 배틀 필드 기본 상태 구현

GyutaeLee committed:
feat : 게임 기본 로직 제작

Parkhyunseo commented on Union/Assets/Scripts/Ability/Ability.cs:
```diff
@@ -0,0 +1,44 @@
+﻿namespace Union.Services.Ability
+{
+    public class Ability
```
Ability보단 Stat이라는 이름이 더 맞는 것 같아요

GyutaeLee replied:
Stat 으로 클래스 네이밍 수정했습니다.

Parkhyunseo commented on Union/Assets/Scripts/Ability/Ability.cs:
```diff
@@ -0,0 +1,44 @@
+        }
+    }
+
+    public class HealthPoint : Ability { }
```
능력별로 클래스를 만들 필요는 없어보여요

GyutaeLee replied:
능력 별로 있던 클래스를 지웠습니다.
Stat 이라는 클래스 자체가 비슷한 성격으로 사용될 가능성이 다분하니, 새로운 함수 추가가 필요하면 Stat 클래스에 추가하는 것으로 합의!

Parkhyunseo commented on Union/Assets/Scripts/Game/BattleField.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    public class BattleField : Singleton<BattleField>
+    {
+        private static class Constants
```
이 방식 어떤 것 같나요?
아직 어색하긴 해서 개선점이 있다면 계속 변경해 나갔으면 합니다.

GyutaeLee replied:
개인 프로젝트에서 사용해보고 있는데 깔끔해서 좋은거 같습니다.
평소에 이곳저곳에 상수를 써두면 관리도 힘들고, 코드 보기도 언짢았는데 아직까진 좋은거 같아요!

Parkhyunseo commented on Union/Assets/Scripts/Game/GameLogic.cs:
```diff
@@ -0,0 +1,81 @@
+
+        private void Initialize()
+        {
+            this._eGameStatus = EGameStatus.None;       
```
상태를 Enum 으로 처리하기보단. 클래스 + 다형성으로 처리하고 FSM처럼 상태가 조건에 따라 또는 호출에 따라 변경되는 식이 되었으면 합니다.

GyutaeLee replied:
게임의 상태를 클래스로 따로 빼내서, 상태에 따라 함수가 각각 호출되도록 구조를 수정했습니다.

이러한 구조를 c#으로 해보는게 처음이라서, 이 부분은 자세히 한 번 더 보고 리뷰 남겨주시면 감사하겠습니다.

Parkhyunseo commented on Union/Assets/Scripts/Game/GamePlayer.cs:
```diff
@@ -0,0 +1,49 @@
+
+        private void OnCollisionEnter(Collision collision)
+        {
+            BattleField.Instance.DecreaseEnemyCount(1); // TO DO : enemy로 주체 이동 필요
```
적의 숫자가 줄어드는 등 사망 처리에 대한 내용은 Enemy쪽에 있는게 맞고, BattleField같은 중간자가 데이터를 가지고 있으려면 이벤트로 처리 하면 좋을 것 같아요

GyutaeLee replied:
아직 Enemy가 없어서 [TO DO] 주석 달아놓고, 추후에 이동하겠습니다.

Parkhyunseo commented on Union/Assets/Scripts/UI/UIGamePlayerStatus.cs:
```diff
@@ -0,0 +1,30 @@
+            this.healthPointImage.fillAmount = 100.0f;
+        }
+
+        public void SetHealthPointUI(int healthPoint)
```
HUD에서 보여주는 플레이어 체력바라면 업데이트 마다 플레이어 체력을 받아와서 보여줘도 괜찮을 것 같아요

GyutaeLee replied:
성능상 참조만 하는 UI들은 Update에서 데이터를 가져와서 처리하도록 수정했습니다.

Parkhyunseo commented on Union/Assets/Scripts/UI/UIGame.cs:
```diff
@@ -0,0 +1,28 @@
+            this.playTimeText.text = text;
+        }
+
+        public void SetEnemyCountText(string text)
```
UI가 시스템을 참조하는게 시스템이 UI를 참조하는 것 보다 괜찮아 보입니다. 그래서 SetEnemyCountText를 호출하지말고 적의 개체수가 변경될 때 발생하는 이벤트를 만들고 그 이벤트에 UI 업데이트 함수를 추가하는 방향을 추천합니다!

JuicyPark replied:
@Parkhyunseo  Update를 최대한 쓰지말자고 생각하면서했는데,  오히려 자주바뀌는 UI는 Update 방식으로 만들어서 디펜던시를 줄인다는 관점은 생각못해본것이라 도움이 되었습니다 :)

GyutaeLee replied:
PlayerStatus.cs 이슈와 동일하게,
성능상 참조만 하는 UI들은 Update에서 데이터를 가져와서 처리하도록 수정했습니다.

Parkhyunseo reviewed (changes requested):
고생하셨습니다~ 리뷰 반영해주시고 추가 질문도 코멘트로 남겨주세요!

JuicyPark commented on Union/Assets/Scripts/Singleton.cs:
```diff
@@ -0,0 +1,24 @@
+using UnityEngine;
```
싱글톤 스크립트의 위치를 Scripts 폴더에 바로 넣어놓으셨는데 다른분들 생각은 어떠신가요?
저는 /Scripts/Singleton/Singleton.cs 로 평소에 올려놓긴하는데 다른분들 의견이 궁금합니다
@Parkhyunseo  @GyutaeLee

GyutaeLee replied:
음.. 이건 테스트용으로 넣은거라, 폴더를 따로 만들지 않았네요. 준수님 말대로 올려도 무방할거 같습니다.
현서님도 동의하시면 폴더 위치 변경하겠습니다. @Parkhyunseo

GyutaeLee replied:
우선 폴더 이동 시켜두었습니다. 추후에 폴더 구조에 대해서도 이야기 해보면 좋을듯 합니다.

JuicyPark reviewed (approved):
현서가 제가할 코멘트까지 올려놔서 저는 바로 Approve로 올려두겠습니다. 고생많으셨습니다

GyutaeLee committed:
- refactor : Ability 관련 클래스
1. Ability -> Stat : 네이밍 수정
2. 불필요한 클래스 제거

GyutaeLee committed:
- fix : Stat 클래스 관련 수정
1. UnitStat : Union.Services.Stat 네임 스페이스로 묶음
2. Ability -> Stat : 누락된 네이밍 변경 수정

GyutaeLee committed:
- feat : Game Logic 상태 구조 변경
1. GameState.cs : 게임 상태가 조건에 따라 호출되었을때 관리하기 편하도록 클래스 분리
2. GameLogic : GameState 구조에 맞게 코드 변경

GyutaeLee committed:
- chore : 사용하지 않는 소스 코드 제거

GyutaeLee committed:
- fix : GameLogic.cs
1. 누락된 null 체크 로직 추가

GyutaeLee committed:
fix : 데이터를 참조만 하는 UI 코드는 Update에서 데이터를 참조하도록 수정

GyutaeLee committed:
feat : 개인 테스트 씬 생성

GyutaeLee committed:
chore : Singleton.cs 폴더 이동

GyutaeLee committed:
fix : SampleScene.unity 컨플릭트 해결

Parkhyunseo reviewed (approved):
굳! 좋습니다. 아직 Stub 오브젝트들이 대부분이라 나머지 구조는 더 개발하면서 보는게 좋을 것 같습니다!
